File: src/lib/richResponse.ts
/**
 * Rich response contract — TypeScript mirror of src-tauri/src/rich_response.rs.
 *
 * The LLM emits `{ parts: MessagePart[] }`. The chat renderer routes
 * each part to its card component by `kind`. Cards are the visual
 * layer over Travis's tool outputs — a map response renders a map,
 * a doc_ref opens the doc viewer, etc.
 *
 * See INTERFACE.md in the cloud repo for the design principles.
 */

export interface RichResponse {
  parts: MessagePart[];
  /** Shell 7 — shape-shifting resume. When Travis wants to bring
   *  archived cards back to the canvas (e.g., user asked "bring back
   *  what I was doing on the CX hire"), include the message IDs here.
   *  The renderer marks them as resurrected in the card lifecycle
   *  store, overriding the 24h archival. */
  resurrect_ids?: string[];
}

export type MessagePart =
  | TextPart
  | MapPart
  | DocRefPart
  | EntityPart
  | CalendarPart
  | T2tConvoPart
  | ActionProposalPart
  | ListPart
  | ChartPart
  | MediaPart
  | ThreadPart
  // v0.28.28 — Phase A rich response types
  | TablePart
  | KeyValuePart
  | CalloutPart
  | QuickReplyPart
  | StepperPart
  | CodeSnippetPart
  | ContactCardPart;

export interface BasePart {
  kind: MessagePart["kind"];
  narration?: string;
  /// v0.28.14 — response channel. When "voice", the frontend speaks
  /// `narration` via Piper. When "chat" (default), text renders only.
  /// When "silent", the part is not rendered — useful for internal
  /// acks. The LLM decides based on ambient/meeting context.
  channel?: "voice" | "chat" | "silent";
}

export interface TextPart extends BasePart {
  kind: "text";
  markdown: string;
}

export interface MapPart extends BasePart {
  kind: "map";
  /// v0.28.2 — route is optional. When the user asks about a PLACE
  /// (e.g. "show me a map of Lagos") the LLM emits a map part with
  /// just `place` populated; when they ask about a ROUTE (A to B)
  /// the LLM emits `route`. Both are valid map surfaces.
  route?: MapRoute;
  place?: MapPlace;
}

export interface MapPlace {
  label: string;
  lat?: number;
  lng?: number;
  region?: string;
  country?: string;
  /// Short human descriptor for the place ("City in Nigeria",
  /// "Neighborhood in Brooklyn").
  descriptor?: string;
}

export interface DocRefPart extends BasePart {
  kind: "doc_ref";
  document_id: number;
  snippet?: string;
}

export interface EntityPart extends BasePart {
  kind: "entity";
  entity_id: number;
  display_name: string;
  facts?: unknown;
}

export interface CalendarPart extends BasePart {
  kind: "calendar";
  window_start: string;
  window_end: string;
  events: CalendarEvent[];
}

export interface T2tConvoPart extends BasePart {
  kind: "t2t_convo";
  query_id: string;
  from_display: string;
  to_display: string;
  question: string;
  drafted_response?: string;
  final_response?: string;
  state: T2tConvoState;
}

export interface ActionProposalPart extends BasePart {
  kind: "action_proposal";
  action_kind: string;
  preview_title: string;
  preview_body: string;
  input: unknown;
}

export interface ListPart extends BasePart {
  kind: "list";
  title: string;
  rows: ListRow[];
}

export interface ChartPart extends BasePart {
  kind: "chart";
  chart_kind: "sparkline" | "bar" | "pie";
  series: ChartSeries[];
}

export interface MediaPart extends BasePart {
  kind: "media";
  url: string;
  media_kind: "image" | "video" | "audio";
  caption?: string;
}

/** A first-class thread. Long-running open-ended context. Collapses
 *  to a summary in the canvas; expands to full scrollable chat with
 *  embedded sub-cards + a thread-local composer. */
export interface ThreadPart extends BasePart {
  kind: "thread";
  thread_id?: string;
  title: string;
  summary?: string;
  turns: ThreadTurn[];
  pinned?: boolean;
}

export interface ThreadTurn {
  author: "user" | "travis";
  parts: MessagePart[];
}

// ─── Phase A (v0.28.28) new part types ──────────────────────────

/** Structured tabular data. Cell values are rendered by column type;
 *  numbers right-align, currency/date formats when set. */
export interface TablePart extends BasePart {
  kind: "table";
  title?: string;
  columns: TableColumn[];
  rows: (string | number | null)[][];
  /** Optional per-row link/action for "click through". */
  row_actions?: RowAction[];
}
export interface TableColumn {
  key: string;
  label: string;
  align?: "left" | "right" | "center";
  format?: "text" | "number" | "currency" | "date" | "duration" | "percent";
  width?: number;
}

/** Compact strip of labeled facts. Better than a Table for a single
 *  entity's attributes. */
export interface KeyValuePart extends BasePart {
  kind: "keyvalue";
  title?: string;
  items: { label: string; value: string; hint?: string }[];
}

/** Semantic message box — info / warn / success / error. */
export interface CalloutPart extends BasePart {
  kind: "callout";
  severity: "info" | "warn" | "success" | "error";
  title?: string;
  body: string;
}

/** Pill options user can click to answer without typing. */
export interface QuickReplyPart extends BasePart {
  kind: "quickreply";
  prompt?: string;
  options: QuickReplyOption[];
}
export interface QuickReplyOption {
  id: string;
  label: string;
  /** When clicked, this string is submitted to Travis as the next
   *  user turn (defaults to label). */
  value?: string;
}

/** Named workflow steps with status. Use for slot-fill progress
 *  ("gathering → drafting → preview → send"). */
export interface StepperPart extends BasePart {
  kind: "stepper";
  title?: string;
  steps: StepperStep[];
}
export interface StepperStep {
  label: string;
  status: "done" | "active" | "pending" | "failed";
  detail?: string;
}

/** Syntax-highlighted code snippet with copy button. */
export interface CodeSnippetPart extends BasePart {
  kind: "code_snippet";
  code: string;
  language?: string;
  filename?: string;
}

/** Person as a first-class card — replaces prose "here's Sarah's info".
 *  Fields mirror the people pack contact schema. */
export interface ContactCardPart extends BasePart {
  kind: "contact_card";
  display_name: string;
  relationship?: string;
  organization?: string;
  email?: string;
  phone?: string;
  birthday?: string;
  notes?: string;
  last_contact_at?: string;
  /** Quick actions the card exposes ("Email", "Log a call"). */
  actions?: RowAction[];
}

// ─── Sub-payloads ─────────────────────────────────────────────────

export interface MapRoute {
  from: LatLng;
  to: LatLng;
  distance_meters: number;
  duration_seconds: number;
  profile?: "driving-car" | "cycling-regular" | "foot-walking";
  destination_label?: string;
  geometry_geojson?: unknown;
}

export interface LatLng {
  lat: number;
  lng: number;
}

export interface CalendarEvent {
  id: string;
  title: string;
  start: string;
  end: string;
  location?: string;
  attendees?: string[];
}

export type T2tConvoState =
  | "sending"
  | "delivered"
  | "considering"
  | "drafted"
  | "answered"
  | "declined";

export interface ListRow {
  id: string;
  label: string;
  meta?: string;
  actions?: RowAction[];
}

export interface RowAction {
  kind: "primary" | "secondary";
  label: string;
  verb: string;
}

export interface ChartSeries {
  label: string;
  points: number[];
}

// ─── Parser ──────────────────────────────────────────────────────

/** Try to parse an LLM response string as a RichResponse. Returns
 *  null if it's not the typed shape — caller should treat as plain
 *  text. */
export function parseRichResponse(raw: string): RichResponse | null {
  try {
    // The LLM may wrap in a ```json fence or leading prose. Try to
    // extract a top-level JSON object first.
    const jsonMatch = raw.match(/\{[\s\S]*\}/);
    if (!jsonMatch) return null;
    const obj = JSON.parse(jsonMatch[0]) as { parts?: unknown };
    if (!Array.isArray(obj.parts)) return null;
    // Cheap validation: every entry has a `kind` string.
    for (const p of obj.parts) {
      if (typeof (p as { kind?: unknown }).kind !== "string") return null;
    }
    return obj as RichResponse;
  } catch {
    return null;
  }
}

/** True if this part kind carries visual content the renderer draws
 *  as a full card (not just a text line). Used to decide narration
 *  playback and layout. */
export function isRichPart(part: MessagePart): boolean {
  return part.kind !== "text";
}

File: src-tauri/src/rich_response.rs
//! Rich response contract — the typed shape the LLM emits.
//!
//! Everything Travis says back to the user is an array of `MessagePart`s.
//! Each part has a `kind` that the renderer switches on to draw the
//! right visual card. Text is a fallback part type; the LLM is prompted
//! to prefer richer types when the answer maps to one.
//!
//! This module owns the ENVELOPE. The individual card payloads
//! (MapCard, DocRefCard, etc.) live near the code that produces them —
//! e.g., MapCard payload lives near the maps tool.
//!
//! See INTERFACE.md in the cloud repo for the design principles.

use serde::{Deserialize, Serialize};

/// The typed reply Travis emits. Always an array of parts, even if
/// there's only one.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RichResponse {
    pub parts: Vec<MessagePart>,
}

/// A single visual/interactive piece the renderer draws. Kind
/// determines which component renders it; payload carries the
/// component-specific data.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum MessagePart {
    /// Prose. Kept for genuinely conversational responses; not the
    /// default. Should carry a short `narration` when speech output
    /// is on.
    Text {
        markdown: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// Inline map card — route, POIs, ETA.
    Map {
        route: MapRoute,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// Reference to a document Travis has. Renders as a doc preview
    /// with an Open button that opens the full doc viewer.
    DocRef {
        document_id: i64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        snippet: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// An entity from Travis's world model — a person, place, thing.
    Entity {
        entity_id: i64,
        display_name: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        facts: Option<serde_json::Value>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// A time-window of calendar events. Renders as a timeline strip.
    Calendar {
        window_start: String, // ISO
        window_end: String,   // ISO
        events: Vec<CalendarEvent>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// A conversation between two Travises (task 311 / T2T rich card).
    T2tConvo {
        query_id: String,
        from_display: String,
        to_display: String,
        question: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        drafted_response: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        final_response: Option<String>,
        state: T2tConvoState,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// A side effect Travis is proposing (send email, book meeting,
    /// invoice generation). Rendered as a card with big Approve /
    /// Edit / Decline controls.
    ActionProposal {
        action_kind: String,
        preview_title: String,
        preview_body: String,
        input: serde_json::Value,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// Structured list of items. Renders as rows with per-row actions.
    List {
        title: String,
        rows: Vec<ListRow>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// A numeric answer with a small chart.
    Chart {
        chart_kind: ChartKind,
        series: Vec<ChartSeries>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },

    /// Image/video preview.
    Media {
        url: String,
        media_kind: MediaKind,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        caption: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        narration: Option<String>,
    },
}

// ─── Sub-payload types ────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MapRoute {
    pub from: LatLng,
    pub to: LatLng,
    pub distance_meters: f64,
    pub duration_seconds: f64,
    #[serde(default)]
    pub profile: Option<String>, // driving-car | cycling-regular | foot-walking
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub destination_label: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub geometry_geojson: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LatLng {
    pub lat: f64,
    pub lng: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CalendarEvent {
    pub id: String,
    pub title: String,
    pub start: String,
    pub end: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub location: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attendees: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum T2tConvoState {
    Sending,
    Delivered,
    Considering,
    Drafted,
    Answered,
    Declined,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListRow {
    pub id: String,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub meta: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<RowAction>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RowAction {
    pub kind: String, // 'primary' | 'secondary'
    pub label: String,
    pub verb: String, // machine name the LLM can call back
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ChartKind {
    Sparkline,
    Bar,
    Pie,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChartSeries {
    pub label: String,
    pub points: Vec<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaKind {
    Image,
    Video,
    Audio,
}

// ─── System prompt fragment (injected into every LLM call) ────────

pub const SYSTEM_PROMPT_FRAGMENT: &str = r#"
## Rich response contract

Your responses to the user are NOT free prose. Emit a JSON object
matching this shape, where `parts` is an array of typed message parts:

```
{ "parts": [ { "kind": "...", ... }, ... ] }
```

Available part kinds and when to use each:

- **map** — the user is asking for directions, wants to see a place, or
  a location matters to the answer. Include distance, duration, route
  geometry when you have it.
- **doc_ref** — the user is asking to see a document Travis has. Emit a
  reference; the client renders the doc viewer.
- **entity** — the answer is about a person, place, or thing the user
  knows. Includes name + structured facts.
- **calendar** — the user is asking about upcoming events or a time
  window.
- **t2t_convo** — a Travis-to-Travis interaction is in progress.
- **action_proposal** — you want to take a side effect (send email,
  invoice generation, book meeting) — emit this and let the user
  approve. Never take the action yourself.
- **list** — the answer is multiple items with common shape.
- **chart** — numeric answer that would read better as a sparkline
  or bar.
- **media** — the answer includes an image, video, or audio clip.
- **text** — a genuinely conversational reply (a joke, small talk,
  a follow-up question, an apology). NOT the default.

Prefer the richest kind that fits. A route request → map, not text.
A doc request → doc_ref, not a text summary. A follow-up "yeah go
ahead" → text, no card needed. Every non-text part MUST carry a
short `narration` string that the voice pipeline can speak — this
is the phone-in-pocket fallback.

Combine parts freely: a route can be a map part + an action_proposal
to add a reminder to leave on time. Keep the array small — 1-3 parts
per reply is typical, 5 is the ceiling.
"#;
